designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,138 @@
+# INuGetProjectServices
+
+* Status: In Review
+* Author(s): [Andy Zivkovic](https://github.com/zivkan)
+* Issue: [1](https://github.com/NuGet/Home/issues/9577) IVsPackageInstallerServices should be async
+
+## Problem Background
+
+The existing IVsPackageInstallerServices service is synchronous, and is being called by several components on the UI thread, causing UI delays in Visual Studio. We are creating new async APIs that will allow us to unblock the UI thread, and use the Visual Studio Service Broker to work in Visual Studio CodeSpaces (and maybe Live Share).
+
+See [NuGet extensibility services](NuGetExtensibilityServices.md) for more details.
+
+## Who are the customers
+
+Visual Studio component and extension developers.
+
+## Goals
+
+* Can be called on any thread (ideally background thread, but work if called on UI thread).
+* Do not block caller (implement async properly, return incomplete task instead).
+* Works with Visual Studio Codespaces, and hopefully Live Share.
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+/// <summary>Service to interact with projects in a solution</summary>
+public interface INuGetProjectServices
+{
+    /// <Summary>Gets the list of packages installed in a project.</Summary>
+    /// <param name="projectId">Project ID (GUID).</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>
+    /// The list of packages in the project.
+    /// </returns>
+    /// <exception cref="System.ArgumentException">When projectId is not a guid.</exception>
+    Task<GetInstalledPackagesResult> GetInstalledPackagesAsync(string projectId, CancellationToken cancellationToken);
+}
+
+/// <summary>Result of a call to INuGetProjectServices.GetInstalledPackagesAsync</summary>
+public sealed class GetInstalledPackagesResult
+{
+    /// <summary>The status of the result</summary>
+    public GetInstalledPackageResultStatus Status { get; }
+
+    /// <summary>List of packages in the project</summary>
+    /// <remarks>May be null if <see cref="Status"/> was not successful</remarks>
+    public IReadOnlyCollection<NuGetInstalledPackage> Packages { get; }
```

Comment:
Given that thius has count, I imagine it doesn't affect performance, but I'm just wondering if we should use IReadOnlyList instead to provide for index access.

Reply:
The RPC contracts best practices document (sadly not public, so I won't link it) says:

> Use `IReadOnlyCollection<T>` for unordered collections or `IReadOnlyList<T>` for ordered collections instead.

Order isn't important for NuGet package definitions, hence the reason I chose that type.